### PR TITLE
Rich-text sanitization #5700

### LINKED ISSUE #5700: Rich-text sanitization
We are going to integrate rich-text support based on HTML, so such content will be stored as a plain HTML. Accordingly, current approach with sanitization of special characters doesn't fit.
Saving plain HTML we need to sanitize it from disallowed content e.g. script tags and such whitelist can be implemented based on OWASP Java HTML Sanitizer

### PR DESCRIPTION
**Pull request by belyabl9**
Fixes #5700

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -50,6 +50,8 @@ dependencies {
                     "org.apache.taglibs:taglibs-standard-impl:1.2.5",
                     "org.jsoup:jsoup:1.9.2"
 
+    compile group: 'com.googlecode.owasp-java-html-sanitizer', name: 'owasp-java-html-sanitizer', version: 'r136'
```
Comment (wkurniawan07) [inline build.gradle]: Let's format this similarly:
`"com.googlecode.owasp-java-html-sanitizer:owasp-java-html-sanitizer:r136"`
and put it on alphabetical order with the other compile-scope dependencies.

Also, I just checked the Maven repository and the latest version of this library is `r239`. Any reason for using `r136`?

File: src/main/java/teammates/common/util/RichTextSecurity.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+                        "br", "ul", "ol", "li")
+                    .allowElements("quote", "ecode")
+                    .toFactory();
+
+    private RichTextSecurity() {
+
+    }
+
+    public static String sanitize(String content) {
+        return policy.sanitize(content);
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/RichTextSecurity.java]: Let's put all these in the `Sanitizer` class, perhaps in a new method called `sanitizeForRichText`. Don't forget to add header comment and unit test for that method.

**Comment by belyabl9** [comment]
Ready to review

File: src/test/java/teammates/test/cases/common/SanitizerTest.java [hunk 3]
```diff
@@ -111,7 +111,13 @@ private void sanitizeHtml_receivesSanitized_returnsUnchanged() {
...
     @Test
     public void testSanitizeForRichText() {
-        // Not tested - using org.apache.commons.lang3.StringEscapeUtils.escapeHtml4()
+        String actualRichText = "<a href=\"https://teammatesv4.appspot.com\" onclick=\"alert('fail');\"></a>"
+                              + "<script>alert('fail');</script>"
+                              + "<p align=\"center\"><strong>Content</strong></p>";
+        String expectedRichText = "<a href=\"https://teammatesv4.appspot.com\"></a>"
+                                + "<p align=\"center\"><strong>Content</strong></p>";
+        String sanitized = Sanitizer.sanitizeForRichText(actualRichText);
+        assertEquals(expectedRichText, sanitized);
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/common/SanitizerTest.java]: Two things here:
- The `+`s should be 2 spaces to the right.
- Add more test cases? In particular, try `null` input, empty string, whitespaces, strings that will stay the same after sanitized, and perhaps 1-2 more examples of how the sanitization works (`onclick` and `script` are demonstrated here, let's demo 1-2 other attributes that will be sanitized).

File: src/main/java/teammates/common/util/Sanitizer.java [hunk 4]
```diff
@@ -114,6 +130,14 @@ public static String sanitizeForJs(String str) {
     }
 
     /**
+     * Sanitizes the string with rich-text. Removes disallowed elements
+     * based on defined policy.
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Sanitizer.java]: Either don't have the line break, or break at "Removes".

File: src/main/java/teammates/common/util/Sanitizer.java [hunk 5]
```diff
@@ -16,7 +19,20 @@
  * and possible threats can be removed first.
  */
 public final class Sanitizer {
-    
+
+    private static PolicyFactory policy = new HtmlPolicyBuilder()
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Sanitizer.java]: See if it looks better if `new HtmlPolicyBuilder()` is on the next line as well (indented 8 spaces w.r.t `private`).

**Comment by wkurniawan07** [comment]
@belyabl9 some changes requested.

**Comment by belyabl9** [comment]
@wkurniawan07 , please, look at changes.

Ready to review

**Comment by wkurniawan07** [comment]
Looks good enough for now, we can always modify it as the need arises :+1:

**Comment by belyabl9** [comment]
Rebased

**Pull request merged**